### ISSUE TITLE
aws ecr describe-images produces inconsistent output

### ISSUE THREAD

**Issue by mackensen** [aws/aws-cli#7411]
### Describe the bug

`aws ecr describe-images` produces inconsistent output depending on whether text or json output is chosen and the repository contains 100 or more tags.

### Expected Behavior

I would expect the command to behave consistently regardless of the output choice.

### Current Behavior

This bug occurs when you have an ECR repository with 100+ tags and you are trying to get the latest tag (see [Stackoverflow](https://stackoverflow.com/questions/43331418/aws-cli-ecr-list-images-get-newest) for related discussion). In this example, there are 124 tags in the repository the "correct" tag (the latest) is `abcd1234`. Here's a query using JSON output:

``` bash
aws --region us-east-1 ecr describe-images --repository-name foo --output json --query 'sort_by(imageDetails,& imagePushedAt)[-1].imageTags[*]'
```

The pagination occurs under the hood and it returns the desired result:

``` json
[
    "abcd1234"
]
```

Same command, using text output:

``` bash
aws --region us-east-1 ecr describe-images --repository-name foo --output text --query 'sort_by(imageDetails,& imagePushedAt)[-1].imageTags[*]'
```

Gives this output:

```
abcd1234
efab5678
```

In other words, the first 100 results get sorted, then the remaining 24 results are sorted, and _both_ are returned.

### Reproduction Steps

Given an ECR repository with over 100 tags, run these two commands:

``` bash
aws --region us-east-1 ecr describe-images --repository-name foo --output json --query 'sort_by(imageDetails,& imagePushedAt)[-1].imageTags[*]'
aws --region us-east-1 ecr describe-images --repository-name foo --output text --query 'sort_by(imageDetails,& imagePushedAt)[-1].imageTags[*]'
```

### CLI version used

2.8.9

### Environment details (OS name and version, etc.)

macOS Monterey 12.6

**Comment by aBurmeseDev** [comment]
Hi @mackensen, thanks for checking in. 

I attempted to replicate the issue but wasn't able to reproduce the same behavior. Could you share the debug logs for both commands by appending `--debug` to each command? That would give us more insight on what the issue is.

**Comment by mackensen** [comment]
@aBurmeseDev sure thing.

- text output: https://gist.github.com/mackensen/445039fd41708f39d1b9dc2c9e97df15
- json output: https://gist.github.com/mackensen/985f7f99f608a625e651c5c682d4b49f

**Comment by aBurmeseDev** [comment]
Thanks for providing the logs. Just wanted to clarify that the issue is with the output result where you see the inconsistency when using different `--output` option but the rest of the command is the same (like `--repository-name`)?

**Comment by mackensen** [comment]
Yes, the inconsistency is between the output types, particularly that text returns two results and json returns one (I assume from internal pagination of some kind).

New gist with both outputs (they're labeled, JSON is first): https://gist.github.com/mackensen/ff6d700da2964464a8cc444ebdfc0793. Please let me know any additional information that you need.

**Comment by aBurmeseDev** [comment]
Hi @mackensen - thank you for your patience. Here's an update followed by the explanation of the behavior after an investigation by the team:

This is an expected behavior and technically not a bug. We currently do not have any plans to change the behavior.
The difference comes from the fact that text outputs are streaming outputs and json outputs are not. When there are multiple pages, streaming outputs will keep streaming a page at a time and return whatever matches the filter rules. Non-streaming outputs on the other hand will concatenate the pages together and then returns whatever matches the filter rules.

So, let’s take the your case for example. Presumably, there are 2 pages’ worth of responses.
The query filter returns the last value in the page, as noted by the `[-1]` bracket: `--query 'sort_by(imageDetails,& imagePushedAt)[-1].imageTags[*]'`
Here’s what’s happening with the `text` output:
1. Stream the first page. Grab the last value in the page `abcd1234`
2. Stream the next page. Grab the last value in the page `efab5678`
3. No more pages. Return [`abcd1234`, `efab5678`]
And here’s what’s happening with the `json` output:
1. Take all 2 pages together and concatenate them
2. Grab the last value in the concatenated page `abcd1234`
3. Return `abcd1234`

Hope it makes sense but please let us know if you have any follow up questions.
